pkg/controller/controlrelation/controller.go

```diff
@@ -0,0 +1,170 @@
+
+	c.Watch(&ConnWatcher{
+		conns: conns,
+	})
+
+	c.Watch(&TopoWatcher{
+		topo:  topo,
+		conns: conns,
+	})
+
+	c.Reconcile(&Reconciler{
+		conns: conns,
+		topo:  topo,
+	})
+
+	return c
+}
+
+// Reconciler reconciles control relations
+type Reconciler struct {
+	conns gnmi.ConnManager
+	topo  topo.Store
+}
+
+// Reconcile reconciles control relations
+func (r *Reconciler) Reconcile(id controller.ID) (controller.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	targetID := id.Value.(topoapi.ID)
+	log.Infof("Reconciling control relation for target %s", targetID)
+	if ok, err := r.reconcileControlRelation(ctx, targetID); err != nil {
+		return controller.Result{}, err
+	} else if ok {
+		return controller.Result{}, nil
+	}
+	return controller.Result{}, nil
+
+}
+
+func (r *Reconciler) reconcileControlRelation(ctx context.Context, targetID topoapi.ID) (bool, error) {
+	conn, err := r.conns.Get(ctx, targetID)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Warnf("Failed to reconcile control relation for target %s, %v", targetID, err)
+			return false, err
+		}
+		// if the connection is not found, find the associated control relation and delete it from topo
+		controlRelationFilter := getControlRelationFilter(utils.GetOnosConfigID(), targetID)
+		controlRelations, err := r.topo.List(ctx, controlRelationFilter)
+		if err != nil {
+			log.Warnf("Failed to reconcile control relation for target %s, %v", targetID, err)
+			return false, err
+		}
+		if len(controlRelations) == 0 {
+			return false, nil
+		}
+
+		return r.reconcileDeleteControlRelation(ctx, &controlRelations[0])
+
+	}
+
+	currentState := conn.State()
+	if currentState == connectivity.Ready {
+		log.Infof("Creating control relation for connection %s", conn.ID())
+		// creates a control relation
+		object := &topoapi.Object{
+			ID:   topoapi.ID(conn.ID()),
+			Type: topoapi.Object_RELATION,
+			Obj: &topoapi.Object_Relation{
+				Relation: &topoapi.Relation{
+					KindID:      topoapi.CONTROLS,
+					SrcEntityID: utils.GetOnosConfigID(),
+					TgtEntityID: conn.TargetID(),
+				},
+			},
+		}
+		err = r.topo.Create(ctx, object)
+		if err != nil {
+			if !errors.IsAlreadyExists(err) {
+				log.Warnf("Failed to reconcile control relation for connection %s, %v", conn.ID(), err)
+				return false, err
+			}
+			return false, nil
+		}
+	} else if currentState == connectivity.TransientFailure || currentState == connectivity.Shutdown {
+		controlRelation, err := r.topo.Get(ctx, topoapi.ID(conn.ID()))
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				log.Warnf("Failed to reconcile control relation for connection %s, %v", conn.ID(), err)
+				return false, err
+			}
+			return false, nil
+		}
+		return r.reconcileDeleteControlRelation(ctx, controlRelation)
+	} else if currentState == connectivity.Idle {
+		log.Debugf("Reconnecting to target:", conn.TargetID())
+		conn.Connect()
```

Comment:
I’m not sure we should really care if the connection state is `IDLE`. It seems like a lot of extra overhead that we can get just by using keep-alive instead. But even without keep-alive, the connection state doesn’t really matter until we are interested in actually sending a request, and the act of sending the request will detect a failure if one has occurred. Mastership can change when that happens. I think we should remove the `Connect` method/calls for now.

Reply:
Removed for now